extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -785,5 +795,53 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		return localRoots;
 	}
 
+	public async undo(): Promise<void> {
+		if (this.undoTocFiles?.length > 0 && this.undoMovedFiles?.length > 0) {
+			let tocFiles: Map<string, tocState> = this.undoTocFiles.pop();
+			let files = this.undoMovedFiles.pop();
+			this.redoMovedFiles.push(files);
+			this.redoTocFiles.push(tocFiles);
+			// restore toc files
+			for (const [tocPath, contents] of tocFiles.entries()) {
+				await this.undoAndRedo(tocPath, contents.previous);
+			}
+			// return files to previous file path
+			for (const [src, dest] of files.entries()) {
+				await fs.move(dest, src);
+			}
+		}
+	}
+
+	public async redo(): Promise<void> {
+		if (this.redoTocFiles?.length > 0 && this.redoMovedFiles?.length > 0) {
+			let tocFiles: Map<string, tocState> = this.redoTocFiles.pop();
+			let files = this.redoMovedFiles.pop();
+			this.undoMovedFiles.push(files);
+			this.undoTocFiles.push(tocFiles);
+			// restore toc files
+			for (const [tocPath, contents] of tocFiles.entries()) {
+				await this.undoAndRedo(tocPath, contents.current);
+			}
+			// return files to previous file path
+			for (const [src, dest] of files.entries()) {
+				await fs.move(src, dest);
+			}
+		}
+	}
+
+	public async undoAndRedo(tocPath: string, toc: string) {
```

Comment:
I'm not sure what this function has to do with undoing/redoing. Could you explain that further?

Reply:
It rewrites the toc and reinitialized the book